### PR TITLE
Make dinov3 ltdetr object detection ONNX/TensorRT export work with FP16

### PR DESCRIPTION
**Pull request by simonschoelly**
## What has changed and why?

This PR fixes two issues with FP16 export to ONNX and TensorRT.

* The export did not work when converting the model with `model.half()`. Instead we always export the model in `fp32` precision and then manually convert the model to `fp16`.
* Converting the model to `fp16` creates overflow/NaN issues around Softmax. We avoid this by carefully using casts to FP32 around these nodes. This previously already worked for TensorRT export, but not for ONNX export.

In addtion
* Removed "auto" as an option for the precision as it was not clear what exactly would happen in that case.
* Exporting to FP16 now requires `simplify=True` - this is, because the simplifier also solves an issue where some two nodes would have the same name in the ONNX Graph.

## How has it been tested?

Added some tests - especially for the removal of unncessary cast nodes. In addition tested on Google Colab and ensured that the FP16 models would correctly run without NaN values. 

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [x] Not needed (internal change)

## Did you update the documentation?

- [ ] Yes
- [x] Not needed (internal change without effects for user)

### CODE REVIEW

File: src/lightly_train/_export/onnx_helpers.py [hunk 1]
```diff
@@ -44,6 +51,75 @@ def precalculate_for_onnx_export() -> Iterator[None]:
...
+            if inp in rewire:
+                node.input[i] = rewire[inp]
+
+    for out in graph.output:
+        if out.name in rewire:
+            out.name = rewire[out.name]
+
+    new_nodes = [n for n in graph.node if id(n) not in nodes_to_remove]
+    del graph.node[:]
+    graph.node.extend(new_nodes)
```
[the remark this answers is not in the document]
Comment (simonschoelly) [reply src/lightly_train/_export/onnx_helpers.py:129]: That would be nice, but makes the solution even more complicated, let's skip if for now.

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py [hunk 2]
```diff
@@ -1111,7 +1112,7 @@ def export_onnx(
         self,
         out: PathLike,
         *,
-        precision: Literal["auto", "fp32", "fp16"] = "auto",
+        precision: Literal["fp32", "fp16"] = "fp32",
         batch_size: int = 1,
```
[the remark this answers is not in the document]
Comment (simonschoelly) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:1114]: It's ok

File: src/lightly_train/_export/onnx_helpers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yutong-xiang-97) [reply src/lightly_train/_export/onnx_helpers.py:129]: I think we can get a relatively easy fix with an `Identity` node. Replace line 118-120 with 

```
output_names = {out.name for out in graph.output}
identity_nodes = []

for output_name, input_name in rewire.items():
    if output_name in output_names:
        identity_nodes.append(
            helper.make_node(
                "Identity",
                inputs=[input_name],
                outputs=[output_name],
                name=f"identity_{output_name}",
            )
        )

graph.node.extend(identity_nodes)
```

[the remark this answers is not in the document]
Comment (simonschoelly) [reply src/lightly_train/_export/onnx_helpers.py:129]: Yes, I know I could do that, but it would make the code even more complicated.

File: src/lightly_train/_export/onnx_helpers.py [hunk 3]
```diff
@@ -44,6 +51,81 @@ def precalculate_for_onnx_export() -> Iterator[None]:
...
+        consumers = input_to_consumers.get(node.output[0], [])
+        for consumer in consumers:
+            if consumer.op_type != "Cast":
+                continue
+            to_attr2 = next((a for a in consumer.attribute if a.name == "to"), None)
+            if to_attr2 is None or to_attr2.i != TensorProto.FLOAT:
+                continue
+
+            # Only remove Cast16 if it has no other consumers.
+            if len(consumers) == 1:
```
Comment (yutong-xiang-97) [inline src/lightly_train/_export/onnx_helpers.py:107]: I don't want to complicate things, but are we sure in this case that each `Cast16` node is bound to have only one consumer?

File: src/lightly_train/_export/onnx_helpers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yutong-xiang-97) [reply src/lightly_train/_export/onnx_helpers.py:129]: If it is exactly what we could do then I don't think it is super complex. Otherwise we are risking tampering the named outputs (like `labels` and `scores` for clsf) which people may use for downstream inference code.

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py [hunk 4]
```diff
@@ -1249,10 +1240,35 @@ def export_onnx(
...
+            from onnxruntime.transformers import float16 as ort_float16
+
+            model_onnx = onnx.load(str(out))
+            # If the input to Softmax are too large the output of Softmax will be NaN values. Therefore we run
+            #  the Softmax computation in fp32. The nodes before Softmax are always MatMul.
+            # TODO (simon, 05/26) Ideally we would only block operators were a Matmul directly feeds into a Softmax.
+            op_block_list = list(ort_float16.DEFAULT_OP_BLOCK_LIST) + [
+                "Softmax",
+                "MatMul",
+            ]
```
[the remark this answers is not in the document]
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:1257]: Do we know how many matmul are there and if we only care about attention-score MatMuls?

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py [hunk 5]
```diff
@@ -1272,10 +1294,15 @@ def export_onnx(
...
+            # Get outputs from the ONNX model. Load from bytes to avoid
+            # ORT errors about missing external data when weights are inline.
+            with open(out, "rb") as f:
+                session = ort.InferenceSession(f.read())
+            onnx_input = dummy_input.cpu()
+            if precision == "fp16":
+                onnx_input = onnx_input.half()
             input_feed = {
-                "images": dummy_input.cpu().numpy(),
+                "images": onnx_input.numpy(),
```
[the remark this answers is not in the document]
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py]: What was the argument that we have this ORT verifier inside `export_onnx`? In Ultralytics there is a separate command `val` for this where you can choose the backend (likely doable for us after we implement the benchmarking command).

File: tests/_task_models/dinov3_ltdetr_object_detection/test_task_model.py [hunk 6]
```diff
@@ -498,3 +498,34 @@ def test_export_onnx__static_batch_size(tmp_path: Path) -> None:
...
+@pytest.mark.skipif(
+    not RequirementCache("onnxruntime"), reason="onnxruntime not installed"
+)
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="Test requires GPU.")
+@pytest.mark.parametrize("decoder_name", ["rtdetrv2", "dfine"])
+def test_export_onnx__fp16(
+    tmp_path: Path, decoder_name: Literal["rtdetrv2", "dfine"]
+) -> None:
+    import onnx
+
```
[the remark this answers is not in the document]
Comment (yutong-xiang-97) [reply tests/_task_models/dinov3_ltdetr_object_detection/test_task_model.py:513]: Related to  https://github.com/lightly-ai/lightly-train/pull/751/#discussion_r3311866725

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py [hunk 7]
```diff
@@ -1386,9 +1413,8 @@ def export_tensorrt(
             max_batchsize=max_batchsize,
             opt_batchsize=opt_batchsize,
             min_batchsize=min_batchsize,
-            # FP32 attention scores required for FP16 model stability. Otherwise output
-            # contains NaN.
-            fp32_attention_scores=True,
+            # We convert the fp32 attention scores already during ONNX export
+            fp32_attention_scores=False,
             verbose=verbose,
```
[the remark this answers is not in the document]
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:1422]: More specifically, the helper calls `export_onnx_fn(**onnx_args)`, and this wrapper never adds `onnx_args["precision"]`, so the ONNX export defaults to fp32 while TensorRT is configured for fp16.

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py [hunk 8]
```diff
@@ -1263,7 +1279,13 @@ def export_onnx(
             import onnx
             import onnxruntime as ort
 
-            onnx.checker.check_model(out, full_check=True)
+            if precision == "fp16" and not torch.cuda.is_available():
+                logger.warning(
+                    "Skipping ONNX full model check for fp16 model because no "
+                    "GPU is available. Run on a GPU to enable full verification."
+                )
+            else:
```
Comment (yutong-xiang-97) [inline src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:1290]: This way `onnx.checker.check_model` is silently skipped for fp16 on CPU. Consider moving it upwards.

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py [hunk 9]
```diff
@@ -1249,10 +1240,35 @@ def export_onnx(
...
+
+            model_onnx = onnx.load(str(out))
+            # If the input to Softmax are too large the output of Softmax will be NaN values. Therefore we run
+            #  the Softmax computation in fp32. The nodes before Softmax are always MatMul.
+            # TODO (simon, 05/26) Ideally we would only block operators were a Matmul directly feeds into a Softmax.
+            op_block_list = list(ort_float16.DEFAULT_OP_BLOCK_LIST) + [
+                "Softmax",
+                "MatMul",
+            ]
+            model_fp16 = ort_float16.convert_float_to_float16(
```
Comment (yutong-xiang-97) [inline src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:1258]: Here the topological order of nodes can be violated. You can add a checker here.

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py [hunk 10]
```diff
@@ -1263,7 +1279,13 @@ def export_onnx(
             import onnx
             import onnxruntime as ort
 
-            onnx.checker.check_model(out, full_check=True)
+            if precision == "fp16" and not torch.cuda.is_available():
```
Comment (yutong-xiang-97) [inline src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py]: There is a `ort.get_available_providers()` for these checks

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py (hunk 9 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:1258]: Maybe also include it in the integration tests

**Review by yutong-xiang-97 — CHANGES_REQUESTED** [review]

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (simonschoelly) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:1257]: We probably only care about attention score MatMuls. I agree that eventually we should have something that only looks at subgraphs of the form MatMul -> Softmax, but the only way to have this maybe automatically with `convert_float_to_float16` is by knowing the names of the nodes.

Anther solution might be quite complicated. So for now we probably should go with what we have here - and then if the performance benchmarks are not good enough, this might be one of the places were we could optimize.

But from Netron, there are definitely more MatMuls than Softmax in the graph.

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py (hunk 10 above)
Comment (simonschoelly) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py]: This is about `onnx.checker.check_model` not about onnxruntime.

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (simonschoelly) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py]: Back when I added the verifier, I think this was just because one might expect that the model works fine when exported, but then creates completely different numerical stuff.

We can definitely think how we can improve this in the future. Not sure if having it as a separate step would make this better though.

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py (hunk 9 above)
Comment (simonschoelly) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:1258]: I am not sure what you are saying here?

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py]: OK I see. I will create a Linear issue for us to look into it later. Personally I would consider abstracting the verification logic to another command and keep the `verify` flag in `export_onnx`.

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (simonschoelly) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:1422]: That is a good point - I will have to test this - I a bit surprised why it did suddenly work for me though.

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py (hunk 9 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:1258]: Sorry I definitely mistyped stuff. A more detailed explanation from Codex:

`ort_float16.convert_float_to_float16(...)` inserts Cast nodes for blocked ops, but some inserted Cast producers end up after the node that consumes them.

Concrete example from the generated graph before onnxslim:

```
node 363: /backbone/rope_embed/Range
  input: /backbone/rope_embed/Range_input_cast_0

node 3979: /backbone/rope_embed/Range_input_cast0 (Cast)
  output: /backbone/rope_embed/Range_input_cast_0
```

That breaks ONNX topological order because `Range` consumes a tensor before the graph has produced it. `remove_redundant_casts()` is not the root cause in this case; the graph is already invalid immediately after `convert_float_to_float16`.

File: src/lightly_train/_export/onnx_helpers.py (hunk 3 above)
Comment (simonschoelly) [reply src/lightly_train/_export/onnx_helpers.py:107]: No it is not, what should happen here if you have something like

```
            .-> Cast32 -> B
A -> Cast16 |
            .-> C
  
```
then we get a graph
```
  .-> B
A |
  .-> Cast16  -> C
```
on the other hand if we only have a single consumer

```
A -> Cast16 -> Cast32 -> B
```
then `Cast16` will be deleted:

```
A -> B
```

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:1257]: Created an linear issue.

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py (hunk 10 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py]: I believe `onnx.checker.check_model` should happen for both cases (hence the comment below). What I meant is to replace `torch.cuda.is_available()` with something like

```
providers = ort.get_available_providers()
if "CUDAExecutionProvider" not in providers:
   ...
```

that is more specific.

Two failure cases

```
- CUDA GPU present but `onnxruntime` (not `onnxruntime-gpu`) installed → `torch.cuda.is_available()` is True, but ORT will fall back to CPU and FP16 inference will likely fail.
- `onnxruntime-gpu` installed in an environment where torch was built CPU-only → the opposite false negative.
``

File: src/lightly_train/_export/onnx_helpers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (simonschoelly) [reply src/lightly_train/_export/onnx_helpers.py:129]: Done

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (simonschoelly) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:1422]: Fixed

File: tests/_task_models/dinov3_ltdetr_object_detection/test_task_model.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (simonschoelly) [reply tests/_task_models/dinov3_ltdetr_object_detection/test_task_model.py:513]: Done

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py (hunk 9 above)
Comment (simonschoelly) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:1258]: Done

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py (hunk 10 above)
Comment (simonschoelly) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py]: Done

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py (hunk 8 above)
Comment (simonschoelly) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:1290]: I think that is ok for now - we can revisit this when we have a working test framework with GPU.

**Review by yutong-xiang-97 — APPROVED** [review]
LGTM!
